File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public bool IsBalanced(TreeNode root) {
        if(root == null) return true;
        if(Math.Abs(depth(root.left) - depth(root.right)) < 2){
            return IsBalanced(root.left) && IsBalanced(root.right);
        }
        else{
            return false;
        }
        
    }


    public int depth(TreeNode p){
        if(p == null) return 0;
        return 1 + Math.Max(depth(p.left), depth(p.right));
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {

        Queue<Integer> queue = new ArrayDeque<>();

        dfs(queue, root);
        int element = -1;
        while(k>0){
            element = queue.poll();
            k--;
        }

        return element;
    }

    private static void dfs(Queue<Integer> queue, TreeNode root) {
        if (root == null) return;

        dfs(queue, root.left);     
        queue.add(root.val);       
        dfs(queue, root.right);
    }
}
